rj_gameplay/rj_gameplay/tactic/wall_tactic.py

```diff
@@ -0,0 +1,191 @@
+from rj_gameplay.skill import move
+import stp.skill as skill
+import numpy as np
+# TODO: replace w/ global param server
+from stp.utils.constants import RobotConstants, BallConstants
+
+class wall_cost(role.CostFn):
+    """Cost function for role request.
+    """
+    def __call__(
+        self,
+        robot: rc.Robot,
+        prev_result: Optional["RoleResult"],
+        world_state: rc.WorldState
+    ) -> float:
+
+        # TODO: make closest robots form wall, rather than setting on init
+        # aka actually use this method
+        # can do this by creating mult cost instances, adding attr like "wall_pt" to this class, 
+        # and giving each roleRequest the correct cost in tactic below
+        return 0.0
+
+def find_wall_pts(num_robots: int, ball_pt: np.ndarray, goal_pt: np.ndarray, world_state: rc.WorldState) -> List[np.ndarray]:
+    """Calculates num_robots points to form a wall between the ball and goal.
+    :return list of wall_pts (as numpy arrays)
+    """
+    # TODO: param server this const
+    WALL_SPACING = BallConstants.RADIUS * 1.9
+
+    # dist is slightly greater than penalty box bounds
+    box_w = world_state.field.penalty_long_dist_m
+    box_h = world_state.field.penalty_short_dist_m
+    line_w = world_state.field.line_width_m
+    DIST_FROM_DEF = RobotConstants.RADIUS + line_w + np.hypot(box_w/2, box_h)
+
+    # check if vision is up and running
+    # (if it is these points should not be equal)
+    if (ball_pt == goal_pt).all():
+        return None
+
+    # get direction vec
+    dir_vec = (ball_pt - goal_pt) / np.linalg.norm(ball_pt - goal_pt)
+    wall_vec = np.array([dir_vec[1], -dir_vec[0]])
+
+    mid_pt = goal_pt + (dir_vec * DIST_FROM_DEF)
+    wall_pts = [mid_pt]
+
+    # set wall points in middle out pattern, given wall dir vector and WALL_SPACING constant
+    wall_pts = [mid_pt]
+    for i in range(num_robots-1):
+        mult = i//2 + 1
+        delta = (mult * (2 * RobotConstants.RADIUS + WALL_SPACING)) * wall_vec 
+        if i % 2: delta = -delta
+        wall_pts.append(mid_pt + delta)
+
+    return wall_pts
+
+def match_robots_to_wall(num_robots: int, ball_pt: np.ndarray, goal_pt: np.ndarray, world_state: rc.WorldState) -> Dict[int, Tuple[rc.Robot, np.ndarray]]:
+    """Matches all wall_points to a robot.
+    :return dictionary of robot_id to (rc.Robot, pt)
+    """
+    # assignments[robot_id] = (rc.Robot, pt)
+    assignments = {}
+
+    wall_pts = find_wall_pts(num_robots, ball_pt, goal_pt, world_state)
+    if not wall_pts: 
+        # sometimes tactic is called before ball_pos is seen
+        return None
+
+    # assign every wall pt a robot to move there
+    for wall_pt in wall_pts:
+        robot = robot_to_wall_pt(wall_pt, world_state, assignments)
+        assignments[robot.id] = (robot, wall_pt)
+
+    return assignments
+
+def robot_to_wall_pt(wall_pt: np.ndarray, world_state: rc.WorldState, assignments: Dict[int, Tuple[rc.Robot, np.ndarray]]) -> rc.Robot:
```

Comment:
This should really be decided by role assignment, you should just produce a cost for each robot, right?

Reply:
will fix